### ISSUE TITLE
Does a memoized function return a new value if called with different arguments?

### ISSUE THREAD

**Issue by callumlocke** [andreypopp/memoize-decorator#1]
If not, could this behaviour be added?

**Comment by andreypopp** [comment]
This is very simple implementation which doesn't allow memorisation for function with arguments. Only zero argument functions are allowed.

**Comment by callumlocke** [comment]
OK. Would you be interested in adding this feature? Or do you think it would be better as a separate module?

`@memoizePerArgs` or something... not sure what the best terminology would be

**Comment by andreypopp** [comment]
Yeah, I'm open for adding this feature to this module. But under a different API.

**Comment by callumlocke** [comment]
do you mean offering two decorators, with separate named exports, like this?

```js
import {memoize, memoizeArgs} from 'memoize-decorator';
```

**Comment by andreypopp** [comment]
I think we should use `WeakMap` where available for a memoization table. Memoization key should be made configurable and should be the explicit choice, e.g. pass key function to decorator when creating. We could provide some sensible default though, like calling `toJSON` and arg tuple.

**Comment by andreypopp** [comment]
Yeah, `memoizeByArgs` or something like that. I'm not a native speaker but `memoizeArg` sounds like you want to memoize args themselves.
